File: src/console/command/parameter/def/ServerParameter.php
<?php

namespace pocketcloud\cloud\console\command\parameter\def;

use pocketcloud\cloud\console\command\parameter\CommandParameter;
use pocketcloud\cloud\console\command\parameter\exception\ArgumentParseException;
use pocketcloud\cloud\server\CloudServer;
use pocketcloud\cloud\server\CloudServerManager;

readonly class ServerParameter extends CommandParameter {

    public function parseValue(string $input): CloudServer {
        if (($server = CloudServerManager::getInstance()->get($input)) !== null) return $server;
        throw new ArgumentParseException();
    }

    public function onTabCompleteMatch(string $currentArg): array {
        return array_filter(array_keys(CloudServerManager::getInstance()->getAll()), fn(string $server) => str_contains(strtolower($server), $currentArg));
    }

    public function getType(): string {
        return "server";
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/console/command/parameter/def/ServerParameter.java
package de.pocketcloud.cloud.console.command.parameter.def;

import de.pocketcloud.api.component.server.ICloudServer;
import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.console.command.exception.ArgumentParseException;
import de.pocketcloud.cloud.console.command.parameter.BaseCommandParameter;

import java.util.List;

public final class ServerParameter extends BaseCommandParameter {

    public ServerParameter(String name, boolean optional) {
        super(name, optional);
    }

    public ServerParameter(String name, boolean optional, String customErrorMessage) {
        super(name, optional, customErrorMessage);
    }

    @Override
    public Object parseValue(String input) throws ArgumentParseException {
        return PocketCloud.instance().servers().get(input).orElseThrow(() -> new ArgumentParseException("Server not found"));
    }

    @Override
    public List<String> onTabCompleteMatch(String currentArg) {
        return PocketCloud.instance().servers().getAll().stream().map(ICloudServer::name).toList();
    }

    @Override
    public String getType() {
        return "server";
    }
}